### PR TITLE
Adding vite vocab plugin package

### PR DESCRIPTION
**Pull request by DanDroryAu**
## Changes

The current JavaScript ecosystem is moving towards using Vite as the bundler more often. For this reason Vocab is experimenting with supporting the Vite system.

This pr adds a new experimental vite plugin to vocab. This plugin transforms compiled vocab index files into files that handle code splitting for the translations. 

It automatically chunks these files into their respective language chunks. This behaviour can be turned off via the plugin config.

## Usage

Default usage

```ts
import { defineConfig } from 'vite';
import vitePluginVocab from '@vocab/vite';
import configFile from './vocab.config.cjs'; // Your local vocab config

export default defineConfig({
  plugins: [vitePluginVocab({
    configFile,
  });
});
```

## Chunk combination

After investigation it is best that the chunk combination is handled by the user rather than the plugin. Changing the config via the `config` lifecycle may have unintended side-effects.

For this reason the decision has been made that end users should handle the chunking strategy. If you want to combine the language chunks from vocab you can use the provided helper function.

To use the plugins chunk combination simply import the `createVocabChunks` function and use it in the `manualChunks` option of your vite config.

```ts
// vite.config.js
import { defineConfig } from 'vite';
import {
  vocabPluginVite,
  createVocabChunks
} from '@vocab/vite';
import configFile from './vocab.config.cjs';

export default defineConfig({
  plugins: [
    vocabPluginVite({
      configFile
    })
  ],
  build: {
    rollupOptions: {
      output: {
        manualChunks: (id, ctx) => {
          // handle your own manual chunks before or after the vocab chunks.
          const languageChunkName = createVocabChunks(
            id,
            ctx
          );
          if (languageChunkName) {
            // vocab has found a language chunk. Either return it or handle it in your own way.
            return languageChunkName;
          }
        }
      }
    }
  }
});
```

### CODE REVIEW

File: packages/vite/src/consts.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const compiledVocabFileFilter = /\.vocab[\\/]index\.(?:ts|js|cjs|mjs)$/;
+export const virtualModuleId = 'vocab-virtual-module';
```
Comment (askoufis) [inline packages/vite/src/consts.ts:2]: IMO we should follow [vite's virtual module convention](https://vite.dev/guide/api-plugin.html#virtual-modules-convention) here and for the `resolveId` and `load` plugin hooks.
```suggestion
export const virtualModuleId = 'virtual:vocab';
```

File: README.md [hunk 2]
```diff
@@ -181,6 +194,86 @@ module.exports = {
...
+  }
+});
+```
+
+#### VocabPluginOptions
+
+```ts
+type VocabPluginOptions = {
+  /**
+   * The Vocab configuration file.
```
[the remark this answers is not in the document]
Comment (DanDroryAu) [reply README.md:269]: I'll refactor this everywhere. It makes more sense that way yes.

File: test-helpers/src/helpers.ts [hunk 3]
```diff
@@ -112,6 +122,95 @@ export const startFixture = (
...
+export const previewViteFixture: FixtureStartFunction = async (
+  fixtureName: FixtureName,
+  options: Options = {},
+): Promise<TestServer> =>
+  new Promise(async (resolve, reject) => {
+    try {
+      await compileFixtureTranslations(fixtureName);
+
+      // eslint-disable-next-line @typescript-eslint/no-require-imports
+      const config = require(`@vocab-fixtures/${fixtureName}/vite.config.js`);
```
[the remark this answers is not in the document]
Comment (DanDroryAu) [reply test-helpers/src/helpers.ts:166]: ditto

File: packages/vite/package.json [hunk 4]
```diff
@@ -0,0 +1,55 @@
+{
+  "name": "@vocab/vite",
+  "version": "1.0.0",
```
Comment (askoufis) [inline packages/vite/package.json:3]: Not that we're going to publish this right now, but would prefer the package start at `0.0.1` to signify that this plugin doesn't have a stable API yet.

Comment (DanDroryAu) [reply packages/vite/package.json:3]: Cool, What's the usual convention when starting packages?

This is v0.u0.p1?

Comment (askoufis) [reply packages/vite/package.json:3]: I guess it depends whether we're confident that we won't want to make breaking changes any time soon. Given that this is the first iteration of the vite plugin, and that we haven't tested it in a real application yet, I'd rather have the flexibility to make breaking changes without bumping the major version.

File: packages/vite/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    name: 'vite-plugin-vocab',
+    resolveId(id) {
+      if (id.includes(virtualModuleId)) {
+        return id;
+      }
+    },
+    load(id) {
+      if (id.includes(virtualModuleId)) {
+        return virtualResourceLoader(id);
+      }
```
Comment (askoufis) [inline packages/vite/src/index.ts:33]: The other part of the [virtual modules convention](https://vite.dev/guide/api-plugin.html#virtual-modules-convention) involves prefixing the id with `\0`. Unsure if this is _really_ worth doing doing, but worth chatting about.

Comment (DanDroryAu) [reply packages/vite/src/index.ts:33]: Yeah I can have a deeper look into this and see how the virtual modules get created and resolved within vite.

**Review by jahredhope — APPROVED** [review]

File: packages/vite/package.json (hunk 4 above)
Comment (DanDroryAu) [inline packages/vite/package.json:3]:
```suggestion
  "version": "0.0.1",
```

**Pull request merged**